src/express-tooling.ts

```diff
@@ -28,82 +29,102 @@ import { safeLogger } from './safe-logger';
 const AUTHORIZATION_HEADER_FIELD_NAME = 'authorization';
 
 /**
- * Returns a function (express middleware) that validates the scopes against the user scopes
- * attached to the request (for example by `handleOAuthRequestMiddleware`).
- * If the the requested scopes are not matched request is rejected (with 403 Forbidden).
+ * Returns a function (express middleware) that validates the scopes against the user scopes.
+ *
+ * The scopes must be attached to the request (for example by `authenticationMiddleware`).
+ * If the requested scopes are not matched the `onAuthorizationFailedHandler` is called.
+ *
+ * The options object can have the following properties:
+ * - optional logger
+ * - onAuthorizationFailedHandler - a customer handler method that
+ *                                  is executed if callee is not authorized,
+ *                                  If not provided, `response.sendStatus(403)` is called.
+ * - precedenceOptions - Let consumers define a handler to overrule scope checking.
+ *                       ⚠️ If precedence function throws or return `false`,
+ *                        the standard scope validation is applied afterwards.
  *
- * ⚠️ If precedence function throws or return `false`, the standard scope validation is applied afterwards.
  *
  * Usage:
  *  app.get('/path', requireScopesMiddleware['scopeA', 'scopeB'], (req, res) => { // Do route work })
  *
  * @param scopes - array of scopes that are needed to access the endpoint
- * @param logger - optional logger
- * @param precedenceOptions - This options let consumers define a way to over rule scope checking. The parameter is optional.
+ * @param options: ScopeMiddlewareOptions
  *
  * @returns { function(any, any, any): undefined }
  */
-type requireScopesMiddleware = (scopes: string[], logger?: Logger, precedenceOptions?: PrecedenceOptions) => RequestHandler;
+type requireScopesMiddleware = (scopes: string[], options?: ScopeMiddlewareOptions) => RequestHandler;
 const requireScopesMiddleware: requireScopesMiddleware =
-  (scopes, logger, precedenceOptions) =>
-    (req: ExtendedRequest, res: Response, next: NextFunction) => {
+  (scopes, options = {}) =>
+    (request: ExtendedRequest, response: Response, nextFunction: NextFunction) => {
+
+      const {
+        logger,
+        onAuthorizationFailedHandler,
+        precedenceOptions
+      } = options;
+
+      const logOrNothing = safeLogger(logger);
+
+      const authorizationFailedHandler =
+        typeof onAuthorizationFailedHandler === 'function' ?
+          (req: Request, res: Response, next: NextFunction, _scopes: string[], _logger: Logger) =>
```

Comment:
maybe a type for this signature, as it is used multiple times?

Reply:
Here we could lower the noise a lot with a type.